### PR TITLE
cpu/stm32f3: optimized isr vector definitions

### PR DESCRIPTION
**Pull request by haukepetersen**
follow up on #7535

### CODE REVIEW

File: cpu/stm32f3/vectors.c [hunk 1]
```diff
@@ -18,362 +18,227 @@
...
+    [17] = isr_dma1_channel7,        /* [17] DMA1 Channel 7 Interrupt */
+    [21] = isr_can_rx1,              /* [21] CAN RX1 Interrupt */
+    [22] = isr_can_sce,              /* [22] CAN SCE Interrupt */
+    [23] = isr_exti,                 /* [23] External Line[9:5] Interrupts */
+    [24] = isr_tim1_brk_tim15,       /* [24] TIM1 Break and TIM15 Interrupts */
+    [25] = isr_tim1_up_tim16,        /* [25] TIM1 Update and TIM16 Interrupts */
+    [26] = isr_tim1_trg_com_tim17,   /* [26] TIM1 Trigger and Commutation and TIM17 Interrupt */
+    [27] = isr_tim1_cc,              /* [27] TIM1 Capture Compare Interrupt */
+    [28] = isr_tim2,                 /* [28] TIM2 global Interrupt */
+    [32] = isr_i2c1_er,              /* [32] I2C1 Error Interrupt */
```
Comment (aabadie) [inline cpu/stm32f3/vectors.c]: This is very strange, some isr are not there any more: 31, 33, 34. There might be a bug in the script that generates the table ?

Comment (smlng) [reply cpu/stm32f3/vectors.c]: don't panic @aabadie 😲  the above are generic ISRs for all F3 MCUs, you'll find your missing 31, 33, 34 and more below in the MCU specific parts.

Comment (smlng) [reply cpu/stm32f3/vectors.c]: mhm, maybe you're still a bit right, I see 34 below but 31, and 33 are now missing? @haukepetersen according to e.g., [stm32f334x8](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f3/include/vendor/stm32f334x8.h#L126) there should be `[31] = isr_i2c1_ev`

Comment (aabadie) [reply cpu/stm32f3/vectors.c]: indeed, I found 34 after commenting but not the others. So there's an issue I think

**Comment by gebart** [comment]
I wrote this script which could be useful for this purpose:
https://gist.github.com/gebart/41d611dfc99d41980cfec4ac89ea08be
It will create a CSV file with a table of all IRQn numbers defined in the CPU header files given on the command line. It seems to work fine for all of the Cortex-M headers in the repo.

CSV of the IRQs of all STM32 in RIOT: https://www.dropbox.com/s/17szxrqsev6pa48/IRQn-RIOT-STM32.csv?dl=0

**Comment by haukepetersen** [comment]
fixed the missing vectors, turned out that my regex for matching the `_IRQn` definition lines was not catching lines with no whitespace between the comment text and the closing `*/` statement...

All definitions should now be included.

File: cpu/stm32f3/vectors.c [hunk 2]
```diff
@@ -27,350 +26,244 @@ void dummy_handler(void) {
...
+    [60] = isr_dma2_channel5,        /* [60] DMA2 Channel 5 global Interrupt */
+    [61] = isr_adc4,                 /* [61] ADC4  global Interrupt */
+    [64] = isr_comp1_2_3,            /* [64] COMP1, COMP2 and COMP3 global Interrupt via EXTI Line21, 22 and 29 */
+    [65] = isr_comp4_5_6,            /* [65] COMP4, COMP5 and COMP6 global Interrupt via EXTI Line30, 31 and 32 */
+    [66] = isr_comp7,                /* [66] COMP7 global Interrupt via EXTI Line33 */
+    [74] = isr_usb_hp,               /* [74] USB High Priority global Interrupt remap */
+    [75] = isr_usb_lp,               /* [75] USB Low Priority global Interrupt  remap */
+    [76] = isr_usbwakeup_rmp,        /* [76] USB Wakeup Interrupt remap */
+#elif defined(CPU_MODEL_STM32F334R8)
+    [18] = isr_adc1_2,               /* [18] ADC1 & ADC2 Interrupts */
```
Comment (aabadie) [inline cpu/stm32f3/vectors.c]: This is present is all F3 flavors, I think it could be moved to the common part.

**Comment by gebart** [comment]
Would it be possible to eliminate some of the model number checks by using the same pattern as the recent ISR vector table changes on Kinetis? https://github.com/RIOT-OS/RIOT/blob/22c52bd3a837127e34f2fcb8d9abb094e1d8f9c7/cpu/kinetis/vectors.c

In short, use the IRQn symbolic names for the array indices and use preprocessor ifdefs to check if the associated hardware module register pointer is defined for the chosen CPU model.

**Comment by gebart** [comment]
The main benefit would be less repeating yourself and no risk of accidentally assigning an isr to the wrong IRQ number, as long as the vendor header is correct for the CPU.

**Comment by haukepetersen** [comment]
Actually I saw that you did that for the `kinetis` and I like it a lot. Should also be easy to do, so will adapt my script and push an update to this PR.

**Comment by haukepetersen** [comment]
adapted the PR. Played around a bit more with vector table generation for the STMs -> turns out they are far from clean throughout all the CPU families, so something like done for the `kinetis` MCUs is not possible, way too many exceptions for the STMs...

**Comment by gebart** [comment]
Did you look through the csv I posted? I have another one for the complete collection of stm32 headers in the Keil uVision CPU support packages. I used that table for the Kinetis CPUs dragging around the columns in Google docs to see better the similarities and differences between models. Btw, doesn't using the IRQn symbolic names as array indices fix a lot of the subtle differences between models? When IRQs are moved in some models to avoid collisions between different peripherals.

**Comment by haukepetersen** [comment]
> . Btw, doesn't using the IRQn symbolic names as array indices fix a lot of the subtle differences between models?

I thought so and played around with this, but the differences are not so easily solvable. Example:
Some STM CPUs have ADC1, ADC2, ADC3 and map them to the same interrupt ADC_IRQn or sometimes ADC1_2_3_IRQn, others of the same group map them to ADC1_IRQn, ADC2_IRQn. Some even split between ADC1_IRQn and ADC2_3_IRQn.
Now the problem is when writing a global vectors file: how to you guard the vector table entries? Just doing
```c
#ifdef ADC1
     [ADC_IRQn] = isr_adc;
#endif
```
Will work for some stms, but will break for others.... Its very very ugly, so I beleave keeping the vector files for each type of stm cpu separated is the way to go here.

**Comment by gebart** [comment]
@haukepetersen I still think it would be better from a maintainability viewpoint to not have too many `#Ifdef CPU_MODEL_xxx` in the vectors but rather just depend on the memory map macros.

From what I can see, the STM32F3's use that kind of IRQ channel combining for the ADC, COMP, and some channels of TIM are combined with DAC IRQ. For the other modules it should be enough with some kind of ifdef for the hardware register (USART, I2C, SPI).
Also, it may be possible to just do
```c
#if defined(ADC1) && defined(ADC2)
[ADC1_2_IRQn] = isr_adc_1_2;
#elif defined (ADC1)
[ADC_IRQn] = isr_adc_1;
#endif
```
and
```c
#if defined(TIM6)
#if defined(DAC1)
[TIM6_DAC_IRQn] = isr_tim6_dac1;
#else
[TIM6_IRQn] = isr_tim6;
#endif
#endif
```
etc. for the peripherals if they are always combined when they exist.

The Kinetis vector uses some compatibility preprocessor definitions at the beginning of the file to rename some IRQ symbols because of differences between the individual vendor headers which reduces the number of special cases like the above.

**Comment by haukepetersen** [comment]
This might be indeed possible, but at least not for me in reasonable time (there are way more important issues to attend...).

I'd say this PR improves the situation already a bit and I would prefer to go with this PR for now until someone takes the time and improves the vector situation even more. What do you say?!

**Comment by haukepetersen** [comment]
rebased and squashed.

**Comment by aabadie** [comment]
@smlng, do you have other comments here ? I think we can merge.

**Comment by smlng** [comment]
@haukepetersen Murdock found one issue, can you have look. Otherwise this seems good to merge.

**Comment by aabadie** [comment]
This is weird, any application builds fine for me with nucleo-f302.

**Comment by PeterKietzmann** [comment]
needs rebase ...

**Comment by smlng** [comment]
@haukepetersen can you rebase? I think we (HAW, @MrKevinWeiss?) should have at least one nucleo f3 for some quick testing

**Comment by smlng** [comment]
ping again, we should get this in as other STMs CPUs have this update already.

**Comment by smlng** [comment]
same as #7687, we should get this in. Ping @haukepetersen

**Comment by smlng** [comment]
needs rebase

**Pull request closed without merging**